lib/sqlalchemy/dialects/postgresql/base.py

```diff
@@ -2041,6 +2041,167 @@ def _on_metadata_drop(self, target, bind, checkfirst=False, **kw):
             self.drop(bind=bind, checkfirst=checkfirst)
 
 
+class Domain(sqltypes.TypeEngine, sqltypes.SchemaType):
+    """Represent the DOMAIN type."""
+
+    __visit_name__ = "domain"
+
+    def __init__(
+        self,
+        name=None,
+        data_type=TEXT,
+        default=None,
+        constraint=None,
+        collation=None,
+        **kw
+    ):
+        """
+        Construct a DOMAIN.
+        """
+        self.name = name
+        if isinstance(data_type, sqltypes.TypeEngine):
+            self.data_type = data_type
+        else:
+            self.data_type = data_type()
+        self.default = default
+        self.constraint = constraint
+        self.collation = collation
+        super(Domain, self).__init__()
+        sqltypes.SchemaType.__init__(
+            self,
+            name=name,
+            schema=kw.pop("schema", None),
+            metadata=kw.pop("metadata", None),
+            inherit_schema=kw.pop("inherit_schema", False),
+            quote=kw.pop("quote", None),
+            _create_events=kw.pop("_create_events", True),
+        )
+
+    def create(self, bind=None, checkfirst=True, **kw):
+        """Emit ``CREATE DOMAIN``
+
+        :param bind: a connectable :class:`_engine.Engine`,
+         :class:`_engine.Connection`, or similar object to emit
+         SQL.
+        :param checkfirst: if ``True``, a query against
+         the PG catalog will be first performed to see
+         if the type does not exist already before
+         creating.
+
+        """
+        # import pdb; pdb.set_trace()
+        bind._run_ddl_visitor(self.DomainGenerator, self, checkfirst=checkfirst)
+
+    def drop(self, bind=None, checkfirst=True, **kw):
+        """Emit ``DROP DOMAIN``
+
+        :param bind: a connectable :class:`_engine.Engine`,
+         :class:`_engine.Connection`, or similar object to emit
+         SQL.
+        :param checkfirst: if ``True``, a query against
+         the PG catalog will be first performed to see
+         if the type actually exists before dropping.
+
+        """
+        bind._run_ddl_visitor(self.DomainDropper, self, checkfirst=checkfirst)
+
+    def _check_for_name_in_memos(self, checkfirst, kw):
```

Comment:
if we go with this, all the code here that's copied from ENUM needs to be factored out into a common base.  I dont want all this duplication.